geps/gep-1897.md

```diff
@@ -16,7 +16,7 @@ so in order to drive resolution this GEP focuses only on this single piece of fu
 1. The solution must satisfy the following use case: the backend pod has its own
```

Comment:
I was thinking of a use-case where a backend (pod/service) is also receiving mesh traffic (in addition to traffic from the gateway) and is configured to only accept mTLS connections. Can we enhance the `TLSConnectionPolicy` struct to add:

- a mode field (to specify mTLS or TLS)
- a field similar to `TrustedCACertRefs` - say `ClientCertRefs` - which the client will use to provide client cert in case of mTLS

So that this API can be directly used for the mesh-case and also cases where the Gateway could or would like to provide a client-cert with an mTLS connection?

Reply:
I think mTLS is intentionally out of scope for this phase of the GEP, but we want to ensure we have room to add it in the future. So it's probably worth sketching out some possible future extensions for mTLS here, but I don't want to block on deciding on a solution here, just as long as we agree that we've left room for a future solution.

Reply:
The mesh case is also intentionally out of scope for this tightly-scoped proposal.  In the case we'd need to do something for mTLS, I would rather propose a new `mTLSConnectionPolicy`.

Reply:
I guess a new mTLSConnectionPolicy type does not sound that bad after all although a single type like TLSConnectionPolicy supporting mTLS is preferable in which case we should make sure it's possible to add it later.
 
I disagree that mTLS is not a service level concern so should be left out. mTLS = TLS + client certificate. If TLS is a service level concern, then so is mTLS. Also mTLS provides client identity which is used in authz and other(?) mesh configs.

Reply:
I think the issue here is one of term overloading.

@sanjaypujare appears to be using `mTLS` to mean "a connection between a Gateway and backend where the Gateway verifies the backend and the backend verifies the Gateway", or in other words a mutual TLS handshake where the certificate management is done by _the owner of the service_, whereas I think @keithmattix and @howardjohn are more thinking about the Service Mesh use of `mTLS` to mean "a mutual auth handshake where the certificate management is performed silently and transparently by infrastructure". In the latter case, it's _important_ that Gateway API constructs are not required, because if they are, then the certificate management is no longer transparent.

The important part here is that this GEP is _not_ about magic infrastructure that automatically upgrades connections to TLS on your behalf, it's about an app owner being able to say that their app _expects_ TLS and _has no magic to make it work transparently_.

This is why, in the Listener case, I've pushed very hard for us to call similar functionality "client certificate configuration" rather than "mTLS", because the use of the acronym can sometimes, depending on who's using it, carry a _lot_ of extra meaning that others may not know is there.

Now, can I see a use for Client Certificate settings in the future in TLSConnectionPolicy? Possibly, but I think it's important that we keep in mind personas here. The whole idea of having mutual authentication of certificates is that it's expected that the two parties spoke _independently_ to the Certificate Authority, and took some steps to prove that they are who they say they are, and then the Certificate Authority provides a method for the two parties to check that the CA that they both trust says the other is okay.

Having the service or backend owner provide the _client certificate_ details completely bypasses this - at that point, there is very little gain in using TLS at all.

For Gateway API to handle the "service/backend owner is doing their own TLS", _and_ "service or backend owner wants to validate the clients connecting to it", it's on the service or backend owner to provide a method for the Gateway owner to retrieve a certificate, and on _us_, as Gateway API designers, to provide a place _on the Gateway_ to configure the client _keypair_ that should be used (since it's both a private and a public key that are required in the client certificate case).

So, what's the summary of this rambling diatribe?
- We've excluded "mTLS" from this precisely because the fact that different people can use it to mean different things means we end up having discussions _exactly like this one, over and over again_.
- This GEP is about _explicit_ TLS configuration, not the _implicit_, _automatic_ TLS configuration you get with a Service Mesh or similar infrastructure tooling. Perhaps we should update some of the preamble to this effect if this makes sense to people other than me, and we can avoid yet _another_ round of this discussion.
- Using an _explicitly configured client certificate_ for the Gateway end of a TLS connection is a reasonable use case, but that _should be configured on the Gateway_, not as part of the TLSConnectionPolicy.

Reply:
Nicely put.  I'll add some of this to the preamble.